plane/src/proxy/proxy_server.rs

```diff
@@ -0,0 +1,279 @@
+use super::{
+    connection_monitor::ConnectionMonitorHandle,
+    request::{get_and_maybe_remove_bearer_token, subdomain_from_host},
+    route_map::RouteMap,
+};
+use crate::{names::Name, protocol::RouteInfo, SERVER_NAME};
+use bytes::Bytes;
+use dynamic_proxy::{
+    body::{simple_empty_body, SimpleBody},
+    hyper::{
+        body::{Body, Incoming},
+        header::{self, HeaderValue},
+        service::Service,
+        Request, Response, StatusCode, Uri,
+    },
+    proxy::ProxyClient,
+    request::MutableRequest,
+};
+use std::{
+    future::{ready, Future},
+    sync::atomic::{AtomicBool, Ordering},
+};
+use std::{pin::Pin, sync::Arc};
+
+pub struct ProxyStateInner {
+    pub route_map: RouteMap,
+    pub proxy_client: ProxyClient,
+    pub monitor: ConnectionMonitorHandle,
+    pub connected: AtomicBool,
+
+    /// If set, the "root" path (/) will redirect to this URL.
+    pub root_redirect_url: Option<String>,
+}
+
+#[derive(Clone)]
+pub struct ProxyState {
+    pub inner: Arc<ProxyStateInner>,
+}
+
+impl Default for ProxyState {
+    fn default() -> Self {
+        Self::new(None)
+    }
+}
+
+impl ProxyState {
+    pub fn new(root_redirect_url: Option<String>) -> Self {
+        let inner = ProxyStateInner {
+            route_map: RouteMap::new(),
+            proxy_client: ProxyClient::new(),
+            monitor: ConnectionMonitorHandle::new(),
+            connected: AtomicBool::new(false),
+            root_redirect_url,
+        };
+
+        Self {
+            inner: Arc::new(inner),
+        }
+    }
+
+    pub fn set_ready(&self, ready: bool) {
+        self.inner.connected.store(ready, Ordering::Relaxed);
+    }
+
+    pub fn is_ready(&self) -> bool {
+        self.inner.connected.load(Ordering::Relaxed)
+    }
+}
+
+impl Service<Request<Incoming>> for ProxyState {
+    type Response = Response<SimpleBody>;
+    type Error = Box<dyn std::error::Error + Send + Sync>;
+    type Future = Pin<
+        Box<
+            dyn Future<
+                    Output = Result<Response<SimpleBody>, Box<dyn std::error::Error + Send + Sync>>,
+                > + Send,
+        >,
+    >;
```

Comment:
I think this was commented elsewhere by @coderabbitai but can we avoid this monstrosity by using the `async-trait` crate?

Reply:
@paulgb making sure you don't miss this

Reply:
This is a hallucination; `async-trait` will work where the future type is inferred by the compiler, but in the case of `Service` we need to provide a specific associated type for `Future`. `async-trait` is not aware of the shape of `Service` so it can't construct that type for us.

The error we get with coderabbit's code is:

```
not all trait items implemented, missing: `Future`
```

Reply:
Ah ok, thanks for that, didn't realize.